core/src/main/scala/kafka/log/LogConfig.scala

```diff
@@ -69,6 +72,9 @@ case class LogConfig(props: java.util.Map[_, _]) extends AbstractConfig(LogConfi
   val minInSyncReplicas = getInt(LogConfig.MinInSyncReplicasProp)
   val compressionType = getString(LogConfig.CompressionTypeProp).toLowerCase
   val preallocate = getBoolean(LogConfig.PreAllocateEnableProp)
+  val messageFormatVersion = ApiVersion(getString(LogConfig.MessageFormatVersionProp))
```

Comment:
That is indeed a caveat. The reason I make message.format.version a topic level config is because it helps to roll out the change. In many cases topics are owned by different applications, so once the application finishes upgrade we can turn on the new message format for them without waiting for others.

One way to enforce this is to add sanity check in both KafkaConfig and TopicConfigHandler. We can check to make sure the message format version config is valid.